server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java

```diff
@@ -356,8 +385,31 @@ private void checkDataTypeConsistency(TSDataType checkedDataType, MeasurementInf
     }
   }
 
-  protected void convertSpecialClauseValues(QueryPlan queryPlan) {
+  protected void convertSpecialClauseValues(QueryPlan queryPlan) throws QueryProcessException {
     if (specialClauseComponent != null) {
+      // record the without null column corresponding index if the plan is not AlignByDevicePlan
+      // otherwise, we simply add all without null column names into AlignByDevicePlan,
+      // and then let itself do the logic of conversion in
+      // `checkWithoutNullColumnValidAlignByDevice` of `QueryOperator.generateAlignByDevicePlan`
+      if (!queryPlan
+          .getPathToIndex()
+          .isEmpty()) { // align by device queryPlan.getPathToIndex() is empty
+        for (Expression expression : specialClauseComponent.getWithoutNullColumns()) {
+          if (queryPlan.getPathToIndex().containsKey(expression.getExpressionString())) {
+            queryPlan.addWithoutNullColumnIndex(
+                queryPlan.getPathToIndex().get(expression.getExpressionString()));
+          } else {
+            throw new QueryProcessException(QueryPlan.WITHOUT_NULL_FILTER_ERROR_MESSAGE);
+          }
+        }
+      } else if (queryPlan instanceof AlignByDevicePlan) {
+        // add without null column
+        AlignByDevicePlan alignByDevicePlan = (AlignByDevicePlan) queryPlan;
+        for (Expression expression : specialClauseComponent.getWithoutNullColumns()) {
+          alignByDevicePlan.addWithoutNullColumns(expression.getExpressionString());
+        }
+      }
```

Comment:
DO NOT use if else here, we have abstracted the QueryPlan well. 
You can add a method `setWithoutNullColumns` in queryplan, and reimplement it in alignByDevicePlan. Then use `queryplan.setWithoutNullColumns(specialClauseComponent)` here.